src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java

```diff
@@ -415,15 +418,31 @@ private boolean isObject(GraphQLType type) {
     }
 
     private List<Object> assertResult(List<MapOrList> parentResults, Object result) {
-        if (result != null && !(result instanceof List)) {
-            throw new BatchAssertionFailed("invalid result from DataFetcher: List expected");
+        result = convertPossibleArray(result);
+        if (result != null && !(result instanceof Iterable)) {
+            throw new BatchAssertionFailed(String.format("BatchedDataFetcher provided an invalid result: Iterable expected but got '%s'. Affected fields are set to null.", result.getClass().getName()));
         }
         @SuppressWarnings("unchecked")
-        List<Object> values = (List<Object>) result;
-        if (values == null || values.size() != parentResults.size()) {
-            throw new BatchAssertionFailed("BatchedDataFetcher provided invalid number of result values. Affected fields are set to null.");
+        Iterable<Object> iterableResult = (Iterable<Object>) result;
+        if (iterableResult == null) {
+            throw new BatchAssertionFailed("BatchedDataFetcher provided a null Iterable of result values. Affected fields are set to null.");
         }
-        return values;
+        long size = iterableResult.spliterator().estimateSize();
```

Comment:
This is not always safe, as `estimateSize()` can return `Long.MAX_VALUE` in certain cases. Since values are anyway copied to a new list few lines below this, you could just check that one after copying.